initial commit, untested

sohailc opened the pull request:
This PR aims to solve the same problem as "driver/Stanford Research sr830 Lock-in amplifier #743" We are opening a new PR because I deleted my original clone of the QCoDeS repo to get rid of unrelated commit messages, 

NB: Please permanently reject PR 743.

jenshnielsen commented on qcodes/instrument_drivers/stanford_research/SR830.py:
```diff
@@ -4,6 +4,59 @@
+from qcodes.instrument.channel import InstrumentChannel, ChannelList
+
+
+class SR830AuxArray(InstrumentChannel):  # We see that InstrumentChannel, ChannelList are something of a misnomer
```
I would call this SR830AuxChannel because thats what it is

sohailc replied:
Ok

jenshnielsen commented on qcodes/instrument_drivers/stanford_research/SR830.py:
```diff
@@ -4,6 +4,59 @@
+                           unit='V')
+
+
+class SR830ChannelArray(InstrumentChannel):
```
same here I would call this SR830Channel

jenshnielsen commented on qcodes/instrument_drivers/stanford_research/SR830.py:
```diff
@@ -4,6 +4,59 @@
+    def __init__(self, parent, name, channel_number):
+        super().__init__(parent, name)
+
+        self._parent = parent
```
This should not be needed as its assigned in the super class

sohailc replied:
Oh yeah, I forgot

jenshnielsen commented on qcodes/instrument_drivers/stanford_research/SR830.py:
```diff
@@ -4,6 +4,59 @@
+            parameter_class=ChannelBuffer
+        )
+
+        def get_parent():
```
A public accessor is probably a good idea. But I would make it a property and add it to the Super class

sohailc commented:
Should I deprecate the old driver and rename this one? Or are you still going to do that?

jenshnielsen commented:
@sohailc Yes please rename this one and copy the old one back in place and add a deprecation similar to what is done with other old drivers

jenshnielsen commented on qcodes/instrument_drivers/stanford_research/SR830.py:
```diff
@@ -154,6 +154,10 @@ class SR830(VisaInstrument):
 
     def __init__(self, name, address, **kwargs):
+
+        DeprecationWarning("This is an old driver an will be removed. Please use the the new driver: "
```
DeprecationWarnings are hidden by default https://docs.python.org/3.6/library/warnings.html#default-warning-filters so we usually use a RunTimeWarbubg or UserWarning to make sure users actually see it

WilliamHPNielsen replied:
A new kind of bug, the RunTimeWarBug! 😄

jenshnielsen replied:
What where I see nothing ;)

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,602 @@
+    The instrument natively supports this in its TRCL call.
+    """
+
+    def __init__(self, name: str, channel_array: 'SR830Channel', channel: int):
```
This will not work. `channel_array` must be renamed to `instrument`, although that is not a very fitting name in this case. We could consider changing that to `parent` in the parameter class.

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,602 @@
+            raise ValueError('Invalid channel specifier. SR830 only has '
+                             'channels 1 and 2.')
+
+        if not isinstance(channel_array, SR830Channel):
```
`channel_array` -> `instrument`

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,602 @@
+                                   'data buffer of one channel.')
+
+        self.channel = channel
+        self._instrument = channel_array.parent
```
`channel_array` -> `instrument`

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,602 @@
+from qcodes.instrument.channel import InstrumentChannel, ChannelList
+
+
+class SR830AuxChannel(InstrumentChannel):  # We see that InstrumentChannel, ChannelList are something of a misnomer
```
The comment here should be removed.

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,602 @@
+                           get_cmd='OEXP? 3',
+                           get_parser=parse_offset_get)
+
+        aux = ChannelList(self, "Aux", SR830AuxChannel, snapshotable=False)
```
If the `ChannelList` is not snapshotable, then the channel should also be added directly to the instrument as a `submodule`. This will also allow for the intuitive `sr.ch1.ratio()` call.

sohailc replied:
I am adding the ChannelList as a submodule. See line 388

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,602 @@
+
+        params = self._instrument.parameters
+        # YES, it should be: "is not 'none'" NOT "is not None"
+        if params['ch{}_ratio'.format(self.channel)].get() is not 'none':
```
Now that the driver uses channels, this line no longer works. It should be something like `self._instrument.submodules['ch{}'.format(self.channel)].ratio.get()`

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,602 @@
+        if params['ch{}_ratio'.format(self.channel)].get() is not 'none':
+            self.unit = '%'
+        else:
+            disp = params['ch{}_display'.format(self.channel)].get()
```
This line also has to be changed to accomodate the restructuring of the driver.

WilliamHPNielsen commented:
@sohailc, in addition to the changes requested above, it would be good to have an updated example notebook on how to use the channelised driver. It can be very similar to the un-channelised one.

WilliamHPNielsen commented:
@sohailc bump, bump, any news on this guy?

sohailc commented:
Currently there are other priorities. I am asked to deliver a live demo of
the WebUI via skype during Microsoft's quarterly meeting in Redmond which
is taking place next week! I will be spending some of my time on that the
coming days

On Thu, Nov 30, 2017 at 5:00 PM, William H.P. Nielsen <
notifications@github.com> wrote:

> @sohailc <https://github.com/sohailc> bump, bump, any news on this guy?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/QCoDeS/Qcodes/pull/811#issuecomment-348232702>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJ-GjGxf2U7JrtVRqS9T8aBH4sL8o4Aoks5s7tEXgaJpZM4QC4Yb>
> .
>

sohailc commented:
@WilliamHPNielsen @jenshnielsen Can we have another look at this driver? I have added tests and removed the bug which prevented the initialization of the driver.

WilliamHPNielsen commented:
@sohailc Somehow the example notebook got "corrupted" so that the docs build fails. You can look into Travis' report for more details (but I think it's difficult to make sense of what's happening). Perhaps it is better to simply make a new example notebook. Perhaps @jenshnielsen can shed more light on this? In any case, that has to be fixed.

I'll do a more thorough test and review now and post the verdict later today.

jenshnielsen commented:
It's probably a merge gone wrong. The json of the notebook it not vaild. I suggest either re executing it locally if possible. Otherwise you can open the notebook in a text editor and copy the input fields to a new one

sohailc commented:
Yeah i saw that it has something to do with the notebook. I will perform
another push today or tomorrow.

On Jan 22, 2018 11:28, "Jens Hedegaard Nielsen" <notifications@github.com>
wrote:

> It's probably a merge gone wrong. The json of the notebook it not vaild. I
> suggest either re executing it locally if possible. Otherwise you can open
> the notebook in a text editor and copy the input fields to a new one
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/QCoDeS/Qcodes/pull/811#issuecomment-359383019>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJ-GjGCT4e7M0yMTBTpJPa6GZkdsfII6ks5tNGLPgaJpZM4QC4Yb>
> .
>

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,628 @@
+        else:
+            self._instrument._buffer2_ready = True
+
+    def get(self) ->np.ndarray:
```
This might as well be `get_raw` to spare us a warning upon initialisation.

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,628 @@
+
+    def _set_input_config(self, s: str) ->int:
+        if s in ['a', 'a-b']:
+            self.sensitivity.set_validator(self._VOLT_ENUM)
```
`set_validator` is deprecated. We should use `self.sensitivity.vals = ` instead.

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,628 @@
+                               'on': 1,
+                           })
+
+        def parse_offset_get(s):
```
Normal `Parameters` should always be scalar (or strings). The `OEXP` command controls both offset and expansion (scale factor) and should be mapped to two different `Parameters`, just like on the front panel. I realise that this PR did not introduce this mistake, but it should correct it.

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,628 @@
+
+        self._channel_number = channel_number
+        self._instrument = instrument.parent
+        self._channel = None  # This is set by the time we call "prepare_buffer_readout"
```
Why is `_channel` not simply assigned here? It confuses both me and `mypy`.

WilliamHPNielsen commented on qcodes/instrument_drivers/stanford_research/SR830_channels.py:
```diff
@@ -0,0 +1,628 @@
+        for param in [self.X, self.Y, self.R, self.sensitivity]:
+            param.unit = unit
+
+    def _get_input_config(self, s: str) ->int:
```
Doesn't this actually return a string?
